### ISSUE TITLE
Is this working for anyone?

### ISSUE THREAD

**Issue by weaselbase1** [arjenvrh/audi_connect_ha#191]
Installed via HACS, and seems able to log in to AudiConnect, but says "Failed to set up" in the Device screen.  Is it working for anyone?

Logs show the following: 
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 390, in async_setup
    result = await component.async_setup_entry(hass, self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/audiconnect/__init__.py", line 111, in async_setup_entry
    return await data.update(utcnow())
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/audiconnect/audi_account.py", line 162, in update
    if not await self.connection.update(None):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/audiconnect/audi_connect_account.py", line 112, in update
    vehicles_response = await self._audi_service.get_vehicle_information()
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/audiconnect/audi_services.py", line 261, in get_vehicle_information
    raise Exception("Invalid json in get_vehicle_information")
Exception: Invalid json in get_vehicle_information

**Comment by agentadam-code** [comment]
Nope. Not worked in a week. Tried multiple forks which are more maintained too and all result in a similar error message. Guess Audi messed with something again.

**Comment by rick-cross** [comment]
Stopped working for me too yesterday (pretty sure yesterday).

**Comment by escorder** [comment]
installed today and is working fine

**Comment by coreywillwhat** [comment]
stopped working a week or so ago and haven't been able to get it back up. Tried deleting the integration and re-adding and still not working.

**Comment by Borsti666** [comment]
Maybe this is the same problem, https://github.com/evcc-io/evcc/issues/8992 VW, Audi, Seat...changed something

**Comment by danantonson** [comment]
> installed today and is working fine

Is this still working for you? I uninstalled and reinstalled and it's still broken.

**Comment by gabrielmc9** [comment]
Audi uses different endpoints for different regions.  When specifying if it's working for you or not, please specify your region.  

I'm in the US, and it hasn't worked for me in a few weeks.  I'm on HA 2023.7.3, HACS 1.32.1.   

I get  the following error:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 390, in async_setup
    result = await component.async_setup_entry(hass, self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/audiconnect/__init__.py", line 111, in async_setup_entry
    return await data.update(utcnow())
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/audiconnect/audi_account.py", line 162, in update
    if not await self.connection.update(None):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/audiconnect/audi_connect_account.py", line 112, in update
    vehicles_response = await self._audi_service.get_vehicle_information()
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/audiconnect/audi_services.py", line 264, in get_vehicle_information
    response.parse(vins["data"])
  File "/config/custom_components/audiconnect/audi_models.py", line 186, in parse
    for item in data.get("userVehicles"):
TypeError: 'NoneType' object is not iterable

**Comment by shogun168** [comment]
Still broken for me, also US region.

**Comment by rosshenning** [comment]
Add another....US region

**Comment by agentadam-code** [comment]
Yeah mine is also logging in no problem when I setup the plug-in but error on start up. I’ve tried the main forks of this plug-in but same issue. Is anyone maintaining updates anymore? Missing my e-tron automations big time.

**Comment by coreywillwhat** [comment]
> Yeah mine is also logging in no problem when I setup the plug-in but error on start up. I’ve tried the main forks of this plug-in but same issue. Is anyone maintaining updates anymore? Missing my e-tron automations big time.

Nope, the original maintainer posted he won't be working on it and we're out of luck if Audi makes any changes. Been posted for a while they're looking for a new maintainer. I feel you though. Getting the car ac going based on calendar events was incredible.

**Comment by agentadam-code** [comment]
Yeah I saw that and was wondering if anyone had hopped in. If I knew anything about reading APIs I’d try it out. For me I have some complex charging automations using solar. I’ll have to dumb them down a lot if no one can fix this.

**Comment by heymoe** [comment]
I am by no means a python expert but I did stay at a Holiday Inn Express :)  

I'm getting the same error in HA as noted by [gabrielmc9](https://github.com/gabrielmc9) above (https://github.com/arjenvrh/audi_connect_ha/issues/191#issuecomment-1655903248) and I'm located in the US.

I added a debug line to audi_services.py at line 262 to print out the contents of the "vins" variable which is passed to the VehiclesResponse() function in audi_models.py which is throwing the "TypeError: 'NoneType' object is not iterable" error and this is what it contained:

```
{
	'errors': [{
			'message': ' Unauthorized',
			'locations': [{
					'line': 2,
					'column': 2
				}
			],
			'path': ['userVehicles'],
			'extensions': {
				'code': 'UNAUTHENTICATED',
				'response': {
					'status': 401,
					'statusText': 'Unauthorized',
					'body': {
						'errorMessage': 'Error in downstream service',
						'vudsErrorCode': 'UNAUTHORIZED'
					}
				}
			}
		}
	],
	'data': {
		'userVehicles': None
	}
}
```

Since "vins" does contain a "data" object, it didn't error out on line 260 in audi_services.py but since there isn't any actual vehicle info in the "userVehicles" object for the for loop on line 182 in audi_models.pl to iterate through, we get the error that causes the addon to fail.

I also confirmed that at least initial when this addon first starts up it does login correctly as my "loggedin" value is set to True which others in previous comments noted as well but based on the error message in the the "vins" variable it would seem that somewhere between the initial login and when get_vehicle_information() gets called our auth session / token is no longer valid or something.

**Comment by jef-pearlman** [comment]
I noticed a new version was pushed out with PR #188, which seems unrelated to this issue.  I'm still getting the error, including after deleting and re-adding my account.  Just wanted to confirm others were having the same experience.  Thanks!

**Comment by heymoe** [comment]
Yep, I installed the new version that got pushed out the other day and still have the issue reported in this thread sadly.

**Comment by agentadam-code** [comment]
Same. That version does not fix this issue.

**Comment by tieskuh** [comment]
Take a look at this fork: https://github.com/cyr-ius/hass-audiconnect
It is well maintained.

**Comment by coreywillwhat** [comment]
> Take a look at this fork: https://github.com/cyr-ius/hass-audiconnect It is well maintained.

Still not working for me unfortunately.

**Comment by Frankster-NL** [comment]
Really don’t understand the problems you are having; I am running the version of Cyrius without any problems for months now…

**Comment by shogun168** [comment]
I'm in the same boat. Original and forked don't work for me. It accepts the credentials and but just says 'Failed setup, will retry' every time. Has done for months unfortunately

**Comment by coreywillwhat** [comment]
> I'm in the same boat. Original and forked don't work for me. It accepts the credentials and but just says 'Failed setup, will retry' every time. Has done for months unfortunately

Mine is doing the same. "Failed setup, will retry" US Region. Here is the log:
`2023-10-22 13:14:19.056 ERROR (MainThread) [custom_components.audiconnect.coordinator] Unexpected error fetching audiconnect data: 'NoneType' object is not iterable
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 290, in _async_refresh
    self.data = await self._async_update_data()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/audiconnect/coordinator.py", line 47, in _async_update_data
    await self.api.async_update()
  File "/usr/local/lib/python3.11/site-packages/audiconnectpy/api.py", line 58, in async_update
    for response in vehicles_response.getr("data.userVehicles", []):
TypeError: 'NoneType' object is not iterable
2023-10-22 13:14:19.060 WARNING (MainThread) [homeassistant.config_entries] Config entry 'Audi connect' for audiconnect integration not ready yet: 'NoneType' object is not iterable; Retrying in background
2023-10-22 13:14:28.303 ERROR (MainThread) [custom_components.audiconnect.coordinator] Unexpected error fetching audiconnect data: 'NoneType' object is not iterable
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 290, in _async_refresh
    self.data = await self._async_update_data()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/audiconnect/coordinator.py", line 47, in _async_update_data
    await self.api.async_update()
  File "/usr/local/lib/python3.11/site-packages/audiconnectpy/api.py", line 58, in async_update
    for response in vehicles_response.getr("data.userVehicles", []):
TypeError: 'NoneType' object is not iterable`

@Frankster-NL are you in the US and Cyrius fork is working fo ryou?

**Comment by danantonson** [comment]
Same story here. I tried the fork and it doesn't work. I'm in the US region.  Does anyone in the US region have this working?

**Comment by dempanator** [comment]
I'm having the same issues. Only showing 2 entities - Lock and Model. Anyone found a solution yet?

**Comment by ridizy** [comment]
I'm in the US and the change in this PR has mostly fixed this integration for me: https://github.com/arjenvrh/audi_connect_ha/pull/203/commits/eaa7bef8ba2bd22739a197ceda568dfa681c44c9

I've lost the device_tracker position entity (which unfortunately I was using for an automation), but it looks like the rest of the entities are working.

**Comment by coreywillwhat** [comment]
Good news! That did it. 
Thank you so much @ridizy.
Wish I could pay someone to maintain this integration.

**Comment by dempanator** [comment]
@ridizy nice work. Doesn't fix a UK integration unfortunately. Don't suppose you have a fix for that too?!

**Comment by ridizy** [comment]
I can't take credit for it.  The PR is from @reisfni.

**Comment by agentadam-code** [comment]
Thanks so much for pointing this out. Changed it and works for me. Most entities are showing.

**Comment by shogun168** [comment]
Holy crap .. +1 that change works for me. Been waiting on a fix for months. Thanks!

**Comment by dempanator** [comment]
> I can't take credit for it. The PR is from @reisfni.

@reisfni - any idea on a UK/EU fix?

**Comment by alderete** [comment]
I applied the patch to modify the URL used in the `audi_services.py` module, as used in the PR from @reisfni. Restarting Home Assistant didn't immediately solve the problem, but looking in my logs, the (new?) error being returned suggested that I might be getting blocked by something stupid: needing to log in manually and accept new terms and conditions.

So, I logged into Audi Connect (https://www.audiusa.com/), clicked through the "updated terms and conditions" message. Logged into the iOS app, did the same. Restarted the integration. 

Success!

Well, mostly. The integration no longer reports an error, and I'm seeing most (all?) of the entities from the integration. But, I'm still getting some errors in my logs that suggest that there are additional URL updates that are needed:

```
2023-11-04 17:31:22.656 ERROR (MainThread) [custom_components.audiconnect.audi_connect_account] Unable to obtain the vehicle shortTerm tripdata of XYZZY: 403, message='', url=URL('https://mal-1a.prd.ece.vwg-connect.com/api/bs/tripstatistics/v1/vehicles/XYZZY/tripdata/shortTerm?type=list&from=1970-01-01T00:00:00Z&to=2023-11-05T02:01:21Z')
2023-11-04 17:31:22.919 ERROR (MainThread) [custom_components.audiconnect.audi_connect_account] Unable to obtain the vehicle longTerm tripdata of XYZZY: 403, message='', url=URL('https://mal-1a.prd.ece.vwg-connect.com/api/bs/tripstatistics/v1/vehicles/XYZZY/tripdata/longTerm?type=list&from=1970-01-01T00:00:00Z&to=2023-11-05T02:01:22Z')
```

Does anyone know if there are additional URL changes that have been discovered, maybe in the forked version of the integration?

**Comment by reisfni** [comment]
> > I can't take credit for it. The PR is from @reisfni.
> 
> @reisfni - any idea on a UK/EU fix?

I would try logging in / refreshing the MyAudi app on a mobile device on your local network and watching DNS requests from that device to see what address(es) are being resolved. It will likely start with "https://app-api".

In fact the [pull request I submitted](https://github.com/arjenvrh/audi_connect_ha/pull/203/files) was a shot in the dark but glad it fixed it for some folks (seems like it was specific to US). It actually quit working on me again a few days later.  I changed line 253 in the audi_services.py to this "https://app-api.my.aoa.audi.com/vgql/v1/graphql" today and got it working again. 

I'm hesitant to submit another pull request as it seems like this one is specific to US/NA (I suspect the "AOA" in the URL is Audi of America) so I wouldn't want it to break the integration for users in other Geos. 

Someone that knows more Python then me could probably write a statement to look at the Region code that is submitted at setup and point to a the right URL depending on what region code is entered.

**Comment by cdnninja** [comment]
Any issues still outstanding.  This is resolved today but wanted to check before we close it.

**Comment by heymoe** [comment]
@cdnninja I'm in the US and I just installed the latest version via HACS and rebooted HA and it looks like everything is working as expected without having to manually apply a custom fix or what not. Thanks!

**Comment by zorwald** [comment]
After I installed the integration initially two weeks ago, it worked seamlessly for my ID3. 
It stopped working last week-ish and I have not been able to get it working again. Does anyone else have the problem as well?

**Comment by agentadam-code** [comment]
Mines still working fine and I rebooted last night and it connected. I’m still on version 1.5 in the US.  I see there has been a few new updates. Try 1.5 or an earlier version?

**Comment by zorwald** [comment]
Funnily enough, after updating to HA 2024.4.3 just now, it works again. I remember updating to 2024.4.2 and it being broken afterward. I did try rebooting and that was no help. So maybe that was it without me realizing it. 
Sorry for any confusion and maybe it helps somebody.
